※ 시간복잡도 vs 공간복잡도

시간복잡도(Time Complexity) : 알고리즘이 실행되는 시간이 입력 크기에 따라 얼마나 증가하는지를 나타낸다.
공간복잡도(Space Complexity) : 알고리즘이 사용하는 메모리의 양이 입력 크기에 따라 얼마나 증가하는지를 나타낸다.

※ 알고리즘과의 관계

좋은 알고리즘은 실행 시간이 짧고(Time), 메모리 사용량도 적은(Space) 알고리즘이다.
하지만 둘은 트레이드오프(Trade-off) 관계인 경우가 많다.
- 메모리를 더 사용하면 실행 시간을 줄일 수 있다.
- 메모리를 아끼면 실행 시간이 늘어날 수 있다.

1. 시간 복잡도

같은 문제라도 알고리즘에 따라 실행 횟수와 수행 시간이 크게 달라짐을 확인한다.(성능 차이를 비교)
시간복잡도(Big-O)가 프로그램 성능을 예측하는 중요한 기준임을 이해한다. 
시간복잡도란 데이터가 많아질수록 연산 횟수가 얼마나 증가하는지를 나타내는 기준이다.

1) 선형 탐색 → 원하는 데이터 찾기

2) 최댓값 찾기 → 반복문으로 값 비교하기 

3) 버블 정렬 → 정렬 알고리즘 이해하기 

4) 실행시간 비교 → 시간복잡도(Big-O)의 필요성을 체감하기

2. 알고리즘

1) 선형 탐색 - 반복 순환

실습 1. 선형 탐색 

주제: 원하는 숫자 찾기

리스트를 처음부터 끝까지 순차적으로 탐색한다.
데이터 개수가 증가할수록 탐색 횟수가 증가함을 확인한다.
데이터가 10배 증가하면 비교 횟수도 최대 10배 증가한다.

학습 개념
- 순차 탐색
- 시간복잡도 : O(n)

In [1]:
numbers = [15, 8, 23, 7, 31, 18]
target = 23

count = 0

for i in range(len(numbers)):
    count += 1
    if numbers[i] == target:
        print("찾은 위치 :", i)
        break # 찾으면 남은 원소는 비교하지않고 break

print("비교 횟수 :", count)

# 데이터를 한 번씩 확인한다.
# 데이터가 n개면 최대 n번 확인한다.
# 그래서 선형(Linear) 시간이라고 부른다.

찾은 위치 : 2
비교 횟수 : 3


2) 최대값 찾기 - 반복 순환

실습 2. 최댓값 찾기 (10분)

주제: 가장 큰 값 찾기

반복문을 이용하여 최댓값을 직접 찾는다.
전체 데이터를 한 번만 순회하는 알고리즘을 구현한다.
최댓값을 찾으려면 모든 데이터를 확인해야 한다. 따라서 이것도 O(n)이다.

학습 개념
- 누적 비교
- 시간복잡도 : O(n)

In [2]:
numbers = [15, 8, 23, 7, 31, 18]

max_num = numbers[0]
count = 0

for n in numbers:
    count += 1
    if n > max_num:
        max_num = n

print("최댓값 :", max_num)
print("비교 횟수 :", count)

# 모든 데이터를 한 번씩 확인해야 한다.
# 중간에 끝낼 수 없다. 원소 끝까지 비교해야 한다. 선형 탐색 처럼 알고리즘 코드 내 break를 하지 않는다.
# 그래서 O(n)

최댓값 : 31
비교 횟수 : 6


3) 버블 정렬 - 중첩 반복 알고리즘

실습 3. 버블 정렬

주제: 숫자 오름차순 정렬

인접한 두 값을 비교하여 교환하는 과정을 구현한다.
데이터 개수가 증가할수록 비교 횟수가 급격히 증가하는 것을 확인한다.

학습 개념
- 중첩 반복문
- 버블 정렬
- 시간복잡도 : O(n²)

In [3]:
numbers = [5, 3, 8, 1, 4]

# 데이터 5개
# 1회전 : 4번 비교
# 2회전 : 3번 비교
# 3회전 : 2번 비교
# 4회전 : 1번 비교

count = 0

for j in range(len(numbers) - 1):
    for i in range(len(numbers) - 1 - j):
        count += 1
        if numbers[i] > numbers[i + 1]:
            numbers[i], numbers[i + 1] = numbers[i + 1], numbers[i] # 우측 패킹 -> 대입 -> 좌측 언패킹
            # temp = numbers[i]
            # numbers[i] = numbers[i+1]
            # numbers[i+1] = temp

print(numbers)
print("비교 횟수 :", count)

[1, 3, 4, 5, 8]
비교 횟수 : 10


- 반복문 안에 반복문이 중첩
- 첫 번째 반복문이 n번, 만약 두 번째 반복문도 n번이면 전체는 n x n이므로 시간복잡도는 O(n²)이 된다
- 버블정렬 알고리즘은 정확히 비교횟수가 O(n²) 증가하지는 않는다. 
- O(n²)는 데이터가 10배 늘어나면 비교횟수는 약 100배 증가한다.
- 시간복잡도는 정확한 횟수를 나타내는 것이 아니라 증가하는 경향을 나타낸다.
- 정확히 몇배가 늘어나는지는 알고리즘에 따라 다르다.
- 같은 O(n²) 알고리즘이라도 실제 비교 횟수와 실행 시간은 서로 다를 수 있다.
- O(n²)는 정확히 몇 배 증가하는지가 아니라, 데이터가 증가할수록 작업량이 대략 n²에 비례하여 증가함을 의미한다.

4) 실행 횟수 비교

같은 데이터 크기에서 O(n)과 O(n²)의 실행 시간을 비교하여 시간복잡도의 차이를 체감해 보자.
'데이터가 많아질수록 알고리즘의 성능 차이가 얼마나 커지는가'를 확인

실습 4. 실행 횟수 비교

주제: 알고리즘 성능 비교

선형 탐색, 최댓값 찾기, 버블 정렬의 반복 횟수를 직접 출력해 비교한다.
time 모듈을 이용하여 실행 시간을 측정하고 입력 크기에 따른 성능 차이를 확인한다.

[1] 실행 시간 측정 방법
- time.time()을 이용하여 프로그램의 실행 시간을 측정하는 방법을 학습한다.
- 알고리즘의 성능을 숫자로 확인하는 방법을 익힌다.

[2] 선형 탐색(O(n))
- 데이터를 처음부터 끝까지 하나씩 확인하는 알고리즘을 이해한다.
- 데이터 개수가 증가하면 실행 시간도 거의 비례하여 증가함을 확인한다.

[3] 버블 정렬(O(n²))
- 중첩 반복문을 사용하는 정렬 알고리즘을 이해한다.
- 데이터 개수가 증가할수록 실행 시간이 급격히 증가함을 확인한다.

[4] 시간복잡도 비교
- 같은 데이터 개수라도 알고리즘에 따라 실행 시간이 크게 달라질 수 있음을 확인한다.
- 시간복잡도(Big-O)가 프로그램의 성능을 예측하는 중요한 기준임을 이해한다.

In [8]:
import random
import time

# 0~9999 사이의 숫자 1000개를 무작위로 생성
numbers = random.sample(range(10000), 1000)

# ===========================
# 선형 탐색 (O(n))
# ===========================

# 실행 시간 측정 시작
start = time.time()

# 마지막 원소를 찾도록 설정(최악의 경우)
target = numbers[-1]

# 처음부터 끝까지 하나씩 비교
for n in numbers:
    if n == target:
        break

# 실행 시간 측정 종료
end = time.time()

print("선형 탐색 :", end - start)

# ===========================
# 버블 정렬 (O(n²))
# ===========================

# 새로운 무작위 데이터 생성
numbers = random.sample(range(10000), 1000)

# 실행 시간 측정 시작
start = time.time()

# 버블 정렬 수행
for j in range(len(numbers) - 1):
    for i in range(len(numbers) - 1 - j):
        if numbers[i] > numbers[i + 1]:
            numbers[i], numbers[i + 1] = numbers[i + 1], numbers[i]

# 실행 시간 측정 종료
end = time.time()

print("버블 정렬 :", end - start)

선형 탐색 : 0.00015282630920410156
버블 정렬 : 0.14721226692199707


- 같은 1,000개의 데이터를 처리했을 때, 선형 탐색은 약 0.00015초, 버블 정렬은 약  0.147초가 걸렸다. 이번 실험에서는 선형 탐색이 약 980배 빠르게 실행되었다.

- 980배라는 수치는 항상 같은 것은 아니다. 컴퓨터 성능, 실행 환경, 데이터의 상태(이미 정렬되어 있는지 등)에 따라 달라질 수 있다. 이 실습에서 중요한 것은 정확히 몇 배 빠른지가 아니라, O(n) 알고리즘이 O(n²) 알고리즘보다 데이터가 많아질수록 훨씬 유리하다는 점이다.

- 선형 탐색은 데이터를 한 번만 훑기 때문에 매우 빠르다. 반면 버블 정렬은 같은 데이터를 여러 번 반복해서 비교하고 교환하므로 훨씬 많은 연산을 수행한다. 그래서 데이터가 많아질수록 실행 시간 차이가 크게 벌어진다. 이것이 시간복잡도 O(n)과 O(n²)의 차이를 보여주는 대표적인 예이다.

In [9]:
data = [] #[[5,5,5], [5,5,5], [5,5,5]]

for i in range(3):
    row = []
    for j in range(3):
        row.append(5)
    data.append(row)

print(data)

[[5, 5, 5], [5, 5, 5], [5, 5, 5]]


In [ ]:
data = [[5 for j in range(3)] for i in range(3)]

print(data)

[[5, 5, 5], [5, 5, 5], [5, 5, 5]]


In [12]:
for row in data:
    for val in row:
        print(val)

5
5
5
5
5
5
5
5
5


In [ ]:
data = [] # [[10, 20, 30], [40, 50, 60], [70, 80, 90]]
count = 1

for i in range(3):
    row = []
    for j in range(3):
        row.append(count * 10)
        count += 1
    data.append(row)

print(data)

for row in data:
    for val in row:
        print(val, end = ", ")

# 게임 / 수학에서 2차원 리스트 좌표가 필요한 경우
for i in range(len(data) * len(data[0])):
    # 0 ->[0,0], 1 ->[0,1], 2 ->[0,2], 3 ->[1,0], 4 ->[1,1], ... 7 ->[2,1], 8 ->[2,2]
    print(i //len(data[0]), i % len(data[0]), data[i // len(data[0])][i % len(data[0])]) 
   

[[10, 20, 30], [40, 50, 60], [70, 80, 90]]
10, 20, 30, 40, 50, 60, 70, 80, 90, 0 0 10
0 1 20
0 2 30
1 0 40
1 1 50
1 2 60
2 0 70
2 1 80
2 2 90


In [28]:
data = [] 
count = 1

for i in range(1000):
    row = []
    for j in range(1000):
        row.append(count * 10)
        count += 1
    data.append(row)

s1 = time.time()

for row in data:
    for val in row:
        pass

s1_end = time.time()
print(f"이중반복문: {s1_end - s1} 초")

# 게임 / 수학에서 2차원 리스트 좌표가 필요한 경우
s2 = time.time()

for i in range(len(data) * len(data[0])):
    pass

s2_end = time.time()
print(f"단일반복문으로 변경 후 : {s2_end - s1} 초")

이중반복문: 0.03678750991821289 초
단일반복문으로 변경 후 : 0.08128786087036133 초
